ring/ring_test.go

```diff
@@ -203,71 +201,96 @@ func TestDoBatch_QuorumError(t *testing.T) {
-		err := runDoBatch()
-		require.Error(t, err)
-		s, ok := status.FromError(err)
-		require.True(t, ok)
-		require.Equal(t, codes.Code(500), s.Code())
+	updateState := func(instanceIDs []string, state InstanceState) {
+		for _, instanceID := range instanceIDs {
+			inst := ring.ringDesc.Ingesters[instanceID]
+			inst.State = state
+			inst.Timestamp = time.Now().Unix()
+			ring.ringDesc.Ingesters[instanceID] = inst
+			ring.updateRingState(ring.ringDesc)
+		}
+	}
+
+	http429Error := httpgrpc.Errorf(429, "Throttling")
+	http500Error := httpgrpc.Errorf(500, "InternalServerError")
+	ringClientError := batchError{isClientError: true}
+	ringServerError := batchError{isClientError: false}
+
+	testCases := map[string]struct {
+		errors               [3]error
+		unavailableInstances []string
+		acceptableOutcomes   []error
+	}{
+		"no error should return no error": {
+			errors:             [3]error{nil, nil, nil},
+			acceptableOutcomes: nil,
+		},
+		"only one HTTP error should return no error": {
+			errors:             [3]error{http500Error, nil, nil},
+			acceptableOutcomes: nil,
+		},
+		"only one ring.Error error should return no error": {
+			errors:             [3]error{ringClientError, nil, nil},
+			acceptableOutcomes: nil,
+		},
+		"2 HTTP 4xx and 1 HTTP 5xx errors should return 4xx": {
+			errors:             [3]error{http429Error, http500Error, http429Error},
+			acceptableOutcomes: []error{http429Error},
+		},
+		"2 HTTP 5xx and 1 HTTP 4xx errors should return 5xx": {
+			errors:             [3]error{http500Error, http429Error, http500Error},
+			acceptableOutcomes: []error{http500Error},
+		},
+		"1 HTTP 4xx, 1 HTTP 5xx and 1 success should return one of the two HTTP errors": {
+			errors:             [3]error{http429Error, http500Error, nil},
+			acceptableOutcomes: []error{http429Error, http500Error},
+		},
+		"1 HTTP error and 1 unhealthy instance should return that error": {
+			errors:               [3]error{http500Error, nil, nil},
+			unavailableInstances: []string{"2"},
+			acceptableOutcomes:   []error{http500Error},
+		},
+		"2 client and 1 server ring.Errors return a client ring.Error": {
+			errors:             [3]error{ringClientError, ringServerError, ringClientError},
+			acceptableOutcomes: []error{ringClientError},
+		},
+		"2 server and 1 client ring.Errors return a server ring.Error": {
+			errors:             [3]error{ringServerError, ringClientError, ringServerError},
+			acceptableOutcomes: []error{ringServerError},
+		},
+		"1 client, 1 server ring.Errors and 1 success return one of the two ring.Errors": {
+			errors:             [3]error{ringClientError, ringServerError, nil},
+			acceptableOutcomes: []error{ringClientError, ringServerError},
+		},
+		"1 ring.Error and 1 unhealthy instance should return that error": {
+			errors:               [3]error{ringClientError, nil, nil},
+			unavailableInstances: []string{"2"},
+			acceptableOutcomes:   []error{ringClientError},
+		},
```

Comment:
should we have a test case with mixed httpgrpc and server/client errors?

Reply:
I have added a couple of tests for this